### PR TITLE
[BEAM-9945] Report data channel progress via a designated counter.

### PR DESCRIPTION
**Pull request by robertwb**
This turned out to make things a lot cleaner IMHO than depending on the (more distant) output PCollection element counts. Also the index here is now the same space as that used for splitting. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @lukecwik

File: model/pipeline/src/main/proto/metrics.proto [hunk 1]
```diff
@@ -292,6 +292,21 @@ message MonitoringInfoSpecs {
         value: "The remaining amount of work for each active element. Each active element represents an independent amount of work not shared with any other active element."
       }]
     }];
+
+    // The (0-based) index of the latest item processed from the data channel.
+    // This gives an indication of the SDKs progress through the data channel,
+    // and is a lower bound on where it is able to split.
+    // For an SDK that processe items sequentially, this is equivalently the
```
Comment (lukecwik) [inline model/pipeline/src/main/proto/metrics.proto:299]:
```suggestion
    // For an SDK that processes items sequentially, this is equivalently the
```

File: model/pipeline/src/main/proto/metrics.proto [hunk 2]
```diff
@@ -292,6 +292,21 @@ message MonitoringInfoSpecs {
...
+    // number of items fully processed (or -1 if reading has not yet started).
```
Comment (lukecwik) [inline model/pipeline/src/main/proto/metrics.proto:300]:
```suggestion
    // number of items fully processed (or -1 if processing has not yet started).
```

File: model/pipeline/src/main/proto/metrics.proto [hunk 3]
```diff
@@ -292,6 +292,21 @@ message MonitoringInfoSpecs {
...
+    DATA_CHANNEL_READ_INDEX = 18 [(monitoring_info_spec) = {
+      urn: "beam:metric:data_channel:read_index:v1",
+      type: "beam:metrics:sum_int64:v1",
+      required_labels: [ "PTRANSFORM" ],
+      annotations: [{
+        key: "description",
+        value: "The number of items processed from a data channel."
```
Comment (lukecwik) [inline model/pipeline/src/main/proto/metrics.proto:307]:
```suggestion
        value: "The read index of the data channel."
```

File: sdks/go/pkg/beam/core/runtime/harness/monitoring.go [hunk 4]
```diff
@@ -265,8 +269,9 @@ func monitoring(p *exec.Plan) ([]*pipepb.MonitoringInfo, map[string][]byte) {
 		if err != nil {
 			panic(err)
 		}
-		payloads[getShortID(metrics.PCollectionLabels(snapshot.PID), urnElementCount)] = payload
 
+    // TODO(BEAM-9934): This metric should account for elements in multiple windows.
```
Comment (lukecwik) [inline sdks/go/pkg/beam/core/runtime/harness/monitoring.go:273]: spacing

Comment (robertwb) [reply sdks/go/pkg/beam/core/runtime/harness/monitoring.go:273]: Ah, tabs.

**Comment by lukecwik** [comment]
It looks like Flink is crashing since MonitoringInfoMetricName isn't seeing a URN. This could be because the Java code is too strict and doesn't ignore unknown metrics.

**Comment by ibzib** [comment]
> It looks like Flink is crashing since MonitoringInfoMetricName isn't seeing a URN. This could be because the Java code is too strict and doesn't ignore unknown metrics.

Luke, do you know which metrics would be unknown and why?

**Comment by lukecwik** [comment]
> > It looks like Flink is crashing since MonitoringInfoMetricName isn't seeing a URN. This could be because the Java code is too strict and doesn't ignore unknown metrics.
> 
> Luke, do you know which metrics would be unknown and why?

Robert did add a new metric URN here as part of this PR. I only quickly glanced at the test to find out that it wasn't a flake and that more changes are needed.

File: sdks/python/apache_beam/metrics/monitoring_infos.py [hunk 5]
```diff
@@ -56,6 +56,8 @@
     [USER_COUNTER_URN, USER_DISTRIBUTION_URN, USER_GAUGE_URN])
 WORK_REMAINING_URN = common_urns.monitoring_info_specs.WORK_REMAINING.spec.urn
 WORK_COMPLETED_URN = common_urns.monitoring_info_specs.WORK_COMPLETED.spec.urn
+DATA_CHANNEL_READ_INDEX = (
+    common_urns.monitoring_info_specs.DATA_CHANNEL_READ_INDEX.urn)
```
Comment (ibzib) [inline sdks/python/apache_beam/metrics/monitoring_infos.py:60]: Looks like you are missing `spec` here.

Comment (lukecwik) [reply sdks/python/apache_beam/metrics/monitoring_infos.py:60]: good find

Comment (robertwb) [reply sdks/python/apache_beam/metrics/monitoring_infos.py:60]: Yeah, I just found that myself... Not sure why it has a urn and a spec.urn.

**Comment by robertwb** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
@ibzib Can you cherrypick this into the 2.21 release

**Comment by ibzib** [comment]
> @ibzib Can you cherrypick this into the 2.21 release

We can't cherry-pick this directly because it depends on https://github.com/apache/beam/pull/11472.

**Comment by lukecwik** [comment]
We only need the Python portion of the change in 2.21, that should remove the need for #11472.

**Pull request merged**